### PR TITLE
Add solution_id to convolution

### PR DESCRIPTION
**Pull request by kahmed10**

### CODE REVIEW

File: src/targets/gpu/convolution.cpp [hunk 1]
```diff
@@ -97,12 +97,44 @@ shape miopen_convolution::compile(context& ctx,
...
+                                                 x_desc.get(),
+                                                 cd.get(),
+                                                 y_desc.get(),
+                                                 solution_count,
+                                                 &solution_count,
+                                                 solutions.data());
+    if(status != miopenStatusSuccess)
+        MIGRAPHX_THROW("MIOpen Convolution: get solution failed");
+
+    solution_id = solutions.front().solution_id;
```
Comment (pfultz2) [inline src/targets/gpu/convolution.cpp:135]: This should be done in the compile function.

**Comment by pfultz2** [comment]
Also, the finalize method should only call `compile` if the solution_id is `0`.

File: src/targets/gpu/include/migraphx/gpu/convolution.hpp [hunk 2]
```diff
@@ -17,19 +17,25 @@ struct miopen_convolution
     shared<convolution_descriptor> cd;
     miopenConvFwdAlgorithm_t algo{};
     miopenHandle_t handle = nullptr;
+    uint64_t solution_id  = 0;
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/convolution.hpp:20]: We can remove `algo` here since we have the solution_id.

Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/convolution.hpp:20]: we still use `algo` for checking fusions when it's using Winograd

File: src/targets/gpu/include/migraphx/gpu/convolution.hpp [hunk 3]
```diff
@@ -17,19 +17,25 @@ struct miopen_convolution
...
+                    f(self.solution_id, "solution_id"));
     }
 
     std::string name() const { return "gpu::convolution"; }
     shape compute_shape(const std::vector<shape>& inputs) const;
     argument
     compute(context& ctx, const shape& output_shape, const std::vector<argument>& args) const;
-    shape compile(context& ctx, const shape& output_shape, std::vector<shape> inputs);
+    shape find(context& ctx, const shape& output_shape, std::vector<shape> inputs);
+    void get_solution(context& ctx, const shape& output_shape, std::vector<shape> inputs);
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/convolution.hpp:38]: This can be removed, We dont define `get_solution` anywhere.

Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/convolution.hpp]: fixed

**Pull request merged**